### ISSUE TITLE
More efficient `CollectionView.reorder()`?

### ISSUE THREAD

**Issue by Seebiscuit** [marionettejs/backbone.marionette#2596]
So, I've been grumbling (#2584) about `CollectionView.reorder` because it immediately defaults to a full Collection\CompositeView render if the filtered, ordered models set is different from the original collection. In words,

```
  reorder: function() {
    var children = this.children;
    var models = this._filteredSortedModels();
    var modelsChanged = _.find(models, function(model) {
      return !children.findByModel(model);
    });

    if (modelsChanged) {
      this.render();
    } else {
      var els = _.map(models, function(model) {
        return children.findByModel(model).el;
      });

     this.triggerMethod('before:reorder');
      this._appendReorderedChildren(els);
      this.triggerMethod('reorder');
    }
  },
```

In my limited understanding the reason we invoke `render` when the filtered set differs from the original collection is because a model may have been silently added to the collection, and, since, a filter is a function, and thus could depend on the order of models, then if the newly sorted filtered set will be different form the rendered children. This results in `children.findByModel(model)` to return `null`, error-ing out on `children.findByModel(model).el` (`el is not a property of undefined`).

So, I propose that, when we encounter a model that does not have a corresponding child view, we create one on the fly:

```
reorder: function() {
   var child, ChildView;
   var children = this.children, models = this._filteredSortedModels();

   var els = _.map(models, _.bind(function(model,index) {
      if (!(child = children.findByModel(model))) {
         ChildView = this.getChildView(child);
         child = this.addChild(model, ChildView, index);
      }
      return child.el;
   }, this));

   this.triggerMethod('before:reorder');
   this._appendReorderedChildren(els);
   this.triggerMethod('reorder');
}
```

Thoughts?

**Comment by ahumphreys87** [comment]
interesting - @Seebiscuit would you be interesting in submitting a PR for this? I wonder if our tests still pass. They should do, I think. 

I wonder what perf gains we get doing this, could you do a comparison maybe?

**Comment by Seebiscuit** [comment]
Sure. I'll give it a shot.

**Comment by ianmstew** [comment]
@Seebiscuit How are things going?  I'm interested in this as well :)

**Comment by ianmstew** [comment]
By the way, this looks related to #2590.  I would be interested to hear your thoughts--perhaps both issues could share pieces of the implementation?

**Comment by Seebiscuit** [comment]
Wow @ianmstew, really cool reference. Yea, I totally agree that this is a common use case. 

In fact, looking at his code more clearly I realize that 

1. My code is problematic: `this.addChild(model, ChildView, index);` will actually attach (pre/append) the view in its parents, which is an unintended and non-perfomant side-effect.

2. @frankcortes 's code can be improved by using @samccone 's method simply re-arranging DOM elements using `$.append`

I'll try to work on a PR for that. 

Potentially, we can handle `resortView` by using `setFilter`, since it will essentially do what reorder does.

```
resortView: function() {
   if (Marionette.getOption(this, 'reorderOnSort')) {
     this.setFilter();
   } else {
     this.render();
   }
}
```

or, to avoid breaking changes we can just alias `reorder` to `setFilter`
```
constructor: function () {
   ...
   this.reorder = this.setFilter;
   ...
},
resortView: function() {
   if (Marionette.getOption(this, 'reorderOnSort')) {
     this.setFilter();
   } else {
     this.render();
   }
}
```

**Comment by Seebiscuit** [comment]
Ok, got it. Simple fix!

```
  _resolveDeltasForFiltering: function (filter) {
    var previousModels = this._filteredSortedModels();
    var child;
    this.filter = filter;
    var models = this._filteredSortedModels();
    var currentIds = _.pluck(models, 'cid');
    var modelMustBeShown;
    // We resolve the deltas in a different way because we have
    // to respect the sorting algorithm.
    _.each(models, function (model, index) {
      if (!(child = this.children.findByModel(model)))
        this._onCollectionAdd(model, this.collection, { at: index });
      else
        this._insertAfter(child, index);   // <-- We now save children that pass the 
                                           //     filter and call _insertAfter
    }, this);
    _.each(previousModels, function (model) {
      modelMustBeShown = _.contains(currentIds, model.cid);
      if (this.children.findByModel(model) && !modelMustBeShown) {
        this._onCollectionRemove(model);
      }
    }, this);
  },
```
By using `_insertAfter` we ensure that we are only *moving* the view, since `_insertAfter` is only a thin wrapper for 
```
this.$el.append(childView.el);
```
Now all we need to do is alias `reorder` to `setFilter` and we're ready to go:

```
constructor: function() {
  ...
  this.reorder = this.setFilter
  ...
},
```
and we remove the `reorder` property from the CollectionView.

**Comment by frankcortes** [comment]
Excellent contribution, @Seebiscuit ! I could add (and review) these changes in my PR, if you agree :+1: 
As a suggestion, if finally we go on with this, maybe `setFilter` and `reorder` should be two alias of a more general function, named `doRenderForDeltas` o something similar. They have the same functionality although they have different semantic after all.

**Comment by Seebiscuit** [comment]
Yes, please, @frankcortes. I would only change the order of the if statement:

```
if ((child = this.children.findByModel(model))) 
   this._insertAfter(child)
else
    this._onCollectionAdd(model, this.collection, { at: index });
```

Just a suggestion. You might bit need to pass index to `_onCollectionAdd` since we're now rendering all the views in order.

**Comment by ianmstew** [comment]
Thanks for collaborating, @Seebiscuit and @frankcortes!  I could see this going into `2.5` (`next`), because we have already built consensus on #2590.  :smile:  Once #2590 is merged, then this issue can be built on top thanks to your shared foundation.

**Comment by Seebiscuit** [comment]
Nooice :metal:
Can't wait to see the commit. If like to run some perf tests on it.

**Comment by frankcortes** [comment]
@Seebiscuit I'm seeing the fact that `reorder` is an extensive case than `filter`. We have to do a deep refactor about how sorting works, and it will take more time :cry:
My suggestion is to publish the #2592 code in the next minor, and apply the sorting changes for a 2.6 or later. Anyway, the issues are related, but they aren't the same thing.
`filter` attribute was just modifying `filter`... `reorder()` modifies how `collection.sort`listener, `reorderOnSort`attribute, `resortView` and `render` works!
What are your thoughts, @ianmstew?

**Comment by paulfalgout** [comment]
@ianmstew @ahumphreys87 so is this resolved in #2592 or is there something still outstanding here?

**Comment by ianmstew** [comment]
@paulfalgout this is not yet resolved. @frankcortes attempted to solve both this and #2592 with a single solution, but it proved to be a rabbit hole.  #2703 is essentially a continuation of this ticket, so it will resolve this one, too--if we get somewhere with it.

**Comment by ahumphreys87** [comment]
closing in favour of #2723
